### DISCUSSION TITLE
JDTLS not working right

### DISCUSSION THREAD

**Discussion by sabinpocris** [joaotavora/eglot#1044] [General]
Hi,
If I am into a Maven project, I get "_file_ is a non-project file, only syntax errors are reported" .

However, if I change the `-data` parameter I get "`class` cannot be resolved to a type", even though the class's implementation is in the same folder.

Googling I saw [another issue like this](https://github.com/mfussenegger/nvim-jdtls/discussions/151#discussioncomment-1395676), maybe using `(java-mode . ("jdtls" "-data" ".jdtls-cache"))` is not the right choice?

What can I do?

I'm using the last `jdtls` snapshot, last `eglot` version and `openjdk 18.`</div>

**Comment by joaotavora** [comment]
> What can I do?

Ask the jdtls folks, maybe

**Comment by rrudakov** [comment]
Is your project a part of Git repository? As far as I understand eglot uses `project.el` to identify a project, so if you don't have a repository your class will be opened as a single file (not as a project).

**Comment by manuel-uberti** [comment]
Can you try using this snippet?
```emacs-lisp
(with-eval-after-load 'eglot
  (let ((cache
         (expand-file-name (md5 (project-root (eglot--current-project)))
                           (locate-user-emacs-file
                            "eglot-eclipse-jdt-cache"))))
    (add-to-list 'eglot-server-programs
                 `(java-mode "jdtls" "-data" ,cache))))
```

**Reply by manuel-uberti** [reply]
@joaotavora Could this issue be somewhat similar to https://github.com/eclipse/eclipse.jdt.ls/pull/2207#issuecomment-1235353109 ?

**Reply by joaotavora** [reply]
I don't know.  Everything can be be everything.  I saw a comment recently here that hinted at a _regression_ in `jdtls`, i.e. that at some point in the past it used to work without being passed `-data`, whereas now it doesn't. This would be consistent with our observations and help cement the case that `jdtls` -- and not Eglot -- should handle this. Can you help me locate that comment?

**Reply by joaotavora** [reply]
Yup, here it is: https://github.com/eclipse/eclipse.jdt.ls/issues/2191#issuecomment-1245437160

**Reply by manuel-uberti** [reply]
I see. I think Eglot should stop guessing the work to be done with `-data`, then, because it seems like with `-data` either within or outside the Java project directory we are ending up with problems. As you said, if it's a regression in `jdtls`, we might be chasing things that are outside the scope of Eglot.

**Reply by sabinpocris** [reply]
Using this snippet and then creating a git repository fixed this for me (works with jdtls 1.15.0 and 1.9).
Removing this snippet I'm getting ".java is a non-project file, only syntax errors are reported".

**Reply by joaotavora** [reply]
You tried two things at once: which one fixed it? What about just creating a git repository?

**Reply by manuel-uberti** [reply]
Yes, that's what Eglot was doing before using a `.jdtls-cache` directory within the Java project directory. However, I still think it should not be up to Eglot to guess what `jdtls` wants to do with this `-data` thing.

**Reply by sabinpocris** [reply]
> You tried two things at once: which one fixed it? What about just creating a git repository?

My bad.
Well, without the snippet, I'm getting that non-project file message.
If I'm into a simple folder with a few java files, the snippet fixes the problem.
However, in the Maven/Gradle project besides the snippet I also need to create a git repo in order for it to work.
